README.md
# NodeJS Typescript Transformer Example Repository

This repository contains a simple example of a NodeJS Typescript Transformer.

## Entrypoint

The `main.ts` file is the entrypoint for the transformer. It contains the following endpoints:

### GET /health

This endpoint is used by the transformer shell to check if the transformer is healthy.

### POST /transform

This endpoint is used by the transformer shell to transform data.

## 📝 Quick start

### 📑 Generate your own repository from this template

<img alt="Create a new repo from the template" src="./.github/readme/create-repo-from-template.gif" width="900" />
<br/>

## 🚀 Deployment

We try to simplify your development experience, by including a pipeline that will automatically test, build and push the
release artifact to the GitHub release. However, if you wish to modify the GitHub Action workflow, or use your own, you
are completely free to do so.

### 🔧 Preparing our pre-made Pipeline

Before you can utilise our pre-built pipeline, you need to authenticate yourself with it. This is done by:

1. generating a personal access token with the permission of creating artifacts.
2. adding a `new repository secret` to your `Settings > Secrets and variables > Actions`, with the name
   of `RELEASE_GITHUB_TOKEN`, and the value of your newly created personal token

You can follow our step-by-step video guide, if you are not accustomed to this process:

[Automating the release process for your Transformer with GitHub Actions](https://youtu.be/QveUgYLJWe8)

### 1️⃣.2️⃣.0️⃣ Release Artifact Versioning

We utilise [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) to automatically manage the
versioning of your released artifacts. In short:

- prefixing your commits with `feat: ` will trigger a minor version (1.**2**.0)
- prefixing your commits with `fix: ` will trigger a patch version (1.0.**3**)

As an example: `feat: splitting email into username and domain`, or `fix: email now splits correctly by @ symbol`

If you need anything more granular, then you can refer to the link above.

### 🔄 Release Process

If using the correct [conventional commit syntax](#120-release-artifact-versioning), then the release process is as
follows:

Anything that gets pushed to the `main` branch, will trigger a pull request; that runs tests to validate the release.
Once the pull request has been merged the release will be published, together with the artifact.

## 💻 Development

We have **2 recommended methods** of working on your transformer.

1. [Develop it directly from GitHub](#option-1-develop-directly-in-github-simple)
2. [Run it locally]()

Your choice may vary depending on your development environment, preference and/or complexity.

### Option 1: Develop directly in GitHub (Simple)

> You can look at how you start the editor directly in your browser [here](https://github.com/github/dev).

Make your changes
and [push them to your repository](https://code.visualstudio.com/docs/sourcecontrol/intro-to-git#_staging-and-committing-code-changes).
**Bear in mind** that you need to use conventional commits to be able to release your transformer. You can read more
about
it under [Deployment](#-deployment).

### Option B: Develop locally (Advanced)

### 1. ⚙️ Installation️

> Prerequisites:
> - NodeJS
> - Docker
>
> And you must clone the project on to your machine

install dependencies:

```bash
yarn install && yarn build
```

run the transformer shell

```bash
docker-compose -f test/docker/docker-compose.yaml up -d
```

### 2. 💻 Development

To start developing with watch mode run:

```bash
yarn build:watch
```

In another terminal or tab, run:

```bash
yarn test:watch
```

to run the tests on the built transformer.

When changes are made any of the files the transformer will be reloaded and the tests will be run again.

> **Note**: The dist directory needs to be writable by the transformer shell.

## 🔎 Development Overview

### Change the transformer

To change the transformer, edit the `transform.entrypoint.ts` file. To add functionality on startup edit
the `start.entrypoint.ts` file. To add additional health checks edit the `health.entrypoint.ts` file.

### Change the input and output data

To change the validation of inputs and outputs edit the `test/expected.json` file. This file specify the event payloads
that are sent to the transformer and the expected output. The `:uuid:` and `:date:` values for the expected outcome
matches to any string.

### Further customization

Change the `test/app.spec.ts` file to add additional tests and more advanced validation. Further change the files in
the `src` directory to add more advanced logic.

## 🌕 Utilise your Transformer in Flowcore

<!-- todo: move this to a another part -->
To use this transformer in the [Flowcore](https://flowcore.io) platform, create a new adapter and point it to the github
release artifact.

The shell will then download the artifact, run it and for each data point post to the `transform` endpoint.

<!-- todo: add a video -->

CHANGELOG.md
# Changelog

## [1.3.1](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.3.0...v1.3.1) (2023-10-04)


### Bug Fixes

* fixed eventid and validtime return ([3872cf7](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/3872cf7ae0fdd518216d549c410d674adab2013b))
* us a non-flowcore related environment variable name ([52114f7](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/52114f7682c02c884edd5dfd415c17bb32b8a8d6))

## [1.3.0](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.2.0...v1.3.0) (2023-09-12)


### Features

* simplified mounting, install production dependencies if package.json is present in dist. ([9b242d5](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/9b242d536aa24ff5eedb00737cf70e1a52bdf384))
* updated docs ([a5303ae](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/a5303aecabf3b68610b739445e846c0f0e211894))


### Bug Fixes

* added docker ignore ([63f4225](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/63f4225e98347906fc93341234e9a3b2e8352bbf))
* added outputs to test ([37fd3f7](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/37fd3f743a5625ddc465a8b670bd3cb898d8e9be))
* added outputs to test ([0d1f3cf](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/0d1f3cfdbf1dec01fbba5ca659e84612628eb760))
* added outputs to test ([602a3c9](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/602a3c936923c35d9511ff5cdaa001a2ee3e66cf))
* added outputs to test ([5302397](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/53023970fa74c9c0e88af5df09a0bf86c41134ec))
* added outputs to test ([87d1c36](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/87d1c364f0a68a9b2339e3e21d887e5b9e5a4cf7))
* bumped transformer shell to 1.4.2 ([eedb1d7](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/eedb1d7f8b4d4e1f746ab83c302c1951670e9d45))
* bumped transformer shell to 1.4.2 ([34b8d9e](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/34b8d9e1ce3a58ed75b38f6269ba3f5598575b74))
* bumped transformer shell to 1.4.3 changed docker compose file ([2ca7b6d](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/2ca7b6d417e427243cb2975ed5bbe81dd4432b62))
* changed permissions on build directory before starting transformer shell ([58edf1c](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/58edf1ce7dfc40db5337388ef581d0645a3dd944))
* set test receiver to use 40xxx range port to avoid conflicts ([9d6fe3e](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/9d6fe3e7ebc3a572a67e59823c0e8102e6a321ab))
* updated readme ([4a7d138](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/4a7d13873fa86d47c8f245ae2277f68751504926))
* updated test pipeline ([e17a051](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/e17a05162c4cebe766192ad59621355470e9c5ee))

## [1.2.0](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.1.0...v1.2.0) (2023-08-31)


### Features

* streamlined DX with JSON configuration ([bd1f9b9](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/bd1f9b94b7452020906f2c89da9d7def20ca551f))

## [1.1.0](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.6...v1.1.0) (2023-08-30)


### Features

* added tests ([ab4c8fc](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/ab4c8fcca45cdcf6445973098a51a1284d8844b7))


### Bug Fixes

* added github host address ([c920023](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/c92002352fcdcbe0b1f493649b8a74c9395dc3b5))
* added tests and lint to workflow ([ff09f9f](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/ff09f9fa8961bb1973388e4157961ec420190738))
* added tests and lint to workflow ([329b13b](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/329b13bcb26c98b89c2f7abb7cb13b64ef3b408c))

## [1.0.6](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.5...v1.0.6) (2023-08-30)


### Bug Fixes

* changed path to include zipped file ([b072fd5](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/b072fd5068975b5349851f5e58e91a6f6f8fd77c))

## [1.0.5](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.4...v1.0.5) (2023-08-30)


### Bug Fixes

* exclude the dist folder path in the zip file ([9d20fc5](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/9d20fc5ded7e3cb878af375ba26ba150954e6748))

## [1.0.4](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.3...v1.0.4) (2023-08-30)


### Bug Fixes

* added production node modules to dist before compressing artifact ([f683906](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/f683906875f9f5e14287116a81cdfa5f51545bda))

## [1.0.3](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.2...v1.0.3) (2023-08-30)


### Bug Fixes

* added write permission ([04a64ef](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/04a64efb0f8d65afde061adf76e491d07fc64b1f))

## [1.0.2](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.1...v1.0.2) (2023-08-30)


### Bug Fixes

* changed upload artifact release ([14b6b5a](https://github.com/flowcore-io/nodejs-typescript-transformer-example/commit/14b6b5a00fd60d9574e771c3dad433e7a08bd819))

## [1.0.1](https://github.com/flowcore-io/nodejs-typescript-transformer-example/compare/v1.0.0...v1.0.1) (2023-08-30)


### Bug Fixes


test/app.spec.ts
import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
import axios from "axios";
import {faker} from "@faker-js/faker";
import {TransformData} from "./fixtures/dtos/transform-data.dto";
import {FLOWCORE_CONSTANT_TIME_BUCKET_FORMAT} from "./fixtures/constants";
import _ from "lodash";
import express from "express";
import {Server} from "http";
import {TransformerBlueprint} from "./fixtures/dtos/transformer-blueprint.dto";
import * as fs from "fs";
import * as path from "path";
import waitForExpect from "wait-for-expect";


dayjs.extend(utc);

const TRANSFORMER_BLUEPRINT: TransformerBlueprint = {
  name: "test-transformer",
  version: "1.0.1",
  runtime: "node",
  artifactUrl: "/app/transformers/test-transformer",
  entrypoint: "main.js",
  startTimeTimeout: 10000,
} as TransformerBlueprint;

const receiverPort = 40300;

describe("NodeJS Test Transformer (e2e)", () => {
  jest.setTimeout(1000000);
  const listeners = new Map<string, jest.MockedFn<any>>();
  const app = express();
  let server: Server;
  let processId: string;

  beforeAll(async () => {
    app.use(express.json());

    app.post("/store/:name", (req, res) => {
      console.log("Received payload", req.params.name, req.body);
      listeners.get(req.params.name)!(req.body);
      res.status(201).send();
    });

    app.get("/health", (req, res) => {
      res.send("OK");
    });

    server = app.listen(receiverPort, () => {
      console.log(`Receiver started on port ${receiverPort}`);
    });

    await waitForExpect(async () => {
      const healthResponse =
        await axios.get(`http://localhost:${receiverPort}/health`);
      expect(healthResponse.status).toEqual(200);
    });

    try {
      const axiosResponse = await axios.post(
        "http://localhost:3001/load",
        TRANSFORMER_BLUEPRINT,
      );

      expect(axiosResponse.data.processId).toBeDefined();

      processId = axiosResponse.data.processId;
    }
    catch (e) {
      console.log("Error loading transformer", e);
    }

  });

  it("should load and process data through a transformer", async () => {
    const receiverName = _.kebabCase(faker.word.noun().toLowerCase());
    const jestFn = jest.fn();
    listeners.set(receiverName, jestFn);

    const expected = JSON.parse(fs.readFileSync(path.join(__dirname, "expected.json"), "utf-8")) as { input: any, output: any }[];

    const definition = {
      hello: "world",
    };

    for (const expectedConfiguration of expected) {
      const eventTime = dayjs.utc(faker.date.recent());
      const data: TransformData = {
        destination: `http://${process.env.HOST_ADDRESS}:${receiverPort}/store/${receiverName}`,
        definition,
        event: {
          eventId: faker.string.uuid(),
          dataCore: faker.string.uuid(),
          eventType: faker.word.noun(),
          aggregator: faker.word.adjective(),
          timeBucket: eventTime.format(FLOWCORE_CONSTANT_TIME_BUCKET_FORMAT),
          validTime: eventTime.toISOString(),
          serializedPayload: JSON.stringify(expectedConfiguration.input),
        },
      };

      const processedResult = await axios.post(
        "http://localhost:3001/transform/" + processId,
        data,
      );

      expect(processedResult.status).toEqual(200);
    }

    for (const expectedConfiguration of expected) {
      const data:{
        [key: string]: any
      } = {};

      for (const key of Object.keys(expectedConfiguration.output)) {
        if (expectedConfiguration.output[key] === ":uuid:") {
          data[key] = expect.any(String);
          continue;
        }

        if (expectedConfiguration.output[key] === ":date:") {
          data[key] = expect.any(String);
          continue;
        }
        data[key] = expectedConfiguration.output[key];
      }

      expect(jestFn.mock.calls).toEqual(expect.arrayContaining([[expect.objectContaining({
        value: data,
      })]]));
    }
  });

  afterAll(async () => {
    console.log("Unloading transformer", processId);
    try {
      await axios.post("http://localhost:3001/unload/" + processId);
    } catch (e) {
      console.log("Error unloading transformer", e);
    }
    server.close();
  });
});


src/main.ts
// -------------------------------------------------------------------
// This is the main entrypoint to the transformer, it should not be
// necessary to modify this file for a basic transformer.
// -------------------------------------------------------------------
import express, {Request, Response} from "express";
import health from "./health.entrypoint";
import transform from "./transform.entrypoint";
import start from "./start.entrypoint";

const app = express();

app.use(express.json());

app.get("/health", async (req: Request, res: Response): Promise<Response> => {
  return res.send(await health());
});

app.post("/transform", async (req: Request, res: Response): Promise<Response> => {
  const { eventId, validTime, payload } = req.body;

  return res.send(await transform({
    eventId,
    validTime,
    payload,
  }));
});

const run = async (): Promise<void> => {
  const port = process.env.PORT || 4000;
  try {
    await start();
    app.listen(port, () => {
      console.log(`Server started on port ${port}`);
    });
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
};

void run();


src/server.ts
import express, { Request, Response } from "express";
import * as console from "console";

const app = express();

app.use(express.json());

app.get("/health", (req: Request, res: Response): Response => {
  // return 200 if the transformer is healthy, required by the transformer shell
  // no transformations will be performed if the transformer is not healthy
  return res.send("OK");
});

app.post("/transform", (req: Request, res: Response): Response => {
  // transform the payload here, return the transformed payload in the response
  const { eventId, validTime, payload } = req.body;

  console.log("Received2 payload", eventId, validTime, payload);

  return res.send({
    eventId,
    validTime,
    ...payload,
  });
});

const start = async (): Promise<void> => {
  const port = process.env.PORT || 4000;
  try {
    app.listen(port, () => {
      console.log(`Server started on port ${port}`);
    });
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
};

void start();
